ddtrace/bootstrap/sitecustomize.py

```diff
@@ -74,6 +69,88 @@ def update_patched_modules():
         EXTRA_PATCHED_MODULES[module] = asbool(should_patch)
 
 
+if PY2:
+    _unloaded_modules = []
+
+
+def gevent_is_installed():
+    # https://stackoverflow.com/a/51491863/735204
+    if sys.version_info >= (3, 4):
+        return importlib.util.find_spec("gevent")
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader("gevent")
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module("gevent")
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module("gevent")
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="auto").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        log.debug("skipping sitecustomize module unload because of DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == 0")
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        log.debug(
+            "skipping sitecustomize module unload because of invalid envvar value"
+            "DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == {}",
+            dd_unload_sitecustomize_modules,
+        )
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not gevent_is_installed():
+        log.debug(
+            "skipping sitecustomize module unload because DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE == auto and "
+            "gevent is not installed"
+        )
+        return False
+    return True
+
+
+def cleanup_loaded_modules():
+    if not should_cleanup_loaded_modules():
+        return
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for m in list(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP):
+        if m.startswith("atexit"):
```

Comment:
I've been meaning to improve these checks with something like `m == module_name or m.startswith("%s." % module_name)`. These current check are vulnerable to modules having the same prefix as the ones we don't want to unload.